James-Lycett/workplace-wellness
# User authentication and some db changes

This contains a lot of major changes so I thought it would be smart to create a separate branch for it. This branch contains all of the changes in pull request #64 so I'm going to close that request to avoid any mishaps that merging both might cause. Additionally, this pull request addresses issues #28 and #66 but also includes some other changes such as the wholesale removal of certain middleware functions that weren't being used.

I understand there's an intimidating number of changes listed below lol it'll probably be way easier for me to run you through all this stuff live rather than you trying to interpret it alone, but you can at least test it on your machine.

**Steps to Test:**
1. Add 'API_SECRET_KEY=296319b8f01c96abd170a5f29f18d288385f3846c956f0ad30c8b089a6029c0b' to backend/.env

2. From backend directory:
- npm i
- npx knex migrate:rollback
- npx knex migrate:latest
- npx knex seed:run
- Keep in mind that this _will_ restructure your databases so if you try to run backend code from another branch it will fail until you rerun the migrations and seeds present in that particular branch

3. Create a new user and/or sign in to a user _passwords for all seeded users are that user's userId_

4. Make sure the dashboard loads up correctly and that everything functions correctly (LOG NEW ACTIVITY, Add a New Employee, "Edit" button on employee cards, and "X" delete employee button)

5. Try accessing a dashboard for a different user (one that isn't logged in) and verify that you are redirected to login page


**Changes:**
_backend_
- Removed extraneous user_health table columns
- Added password_hash column for use in authentication
- Modified seeds accordingly, added a couple of nifty little functions to do so (./backend/src/db/utils)
- Added backend/src/authentication functions
- Added a check condition to the appropriate middleware functions to ensure the user requesting stuff has access to the stuff, or is an admin if it's stuff that an admin needs access to (data.controller update() is as good an example as any, if you wanna see what I mean)
- data.controller create() now takes a password, encrypts it, and sends it to db password_hash column along with the rest of the user object, also has a function "duplicateUsernameExists" to make sure that username isn't taken already
- Removed unused list all entries functions
- Removed unused readByUsername logic in data.controller healthDataExists()
- Updated validateInput() in data.controller to remove validation for the table columns I removed
- Updated validateInput() in entries.controller to actually validate bmi_category (it was commented out and broken, fixed it)
- Changed cors policy to accept credentials headers

_frontend_

- Changed /admin/.../home and /user/.../home routes to a single /dashboard/:userId route
- Accordingly changed routes that LoginPage and CreateAccount navigate to onSubmit
- Added userLogin() to api.js and added some logic to the fetch() function necessary for sending auth token with requests
- Swapped out LoginPage handleSubmit() api call "readByUsername" for "userLogin"
- Added some fields to "Add New Employee" modal form
- CreateAccount page changed up the form and cleaned up the logic a whole lot
- Added ErrorAlert to forms where the user might need to know what went wrong (e.g., "Invalid Username or Password" on login page, "Username already exists" on create account page...)
